ckanext/scheming/validation.py

```diff
@@ -82,6 +85,59 @@ def validator(key, data, errors, context):
 
     return validator
 
+@scheming_validator
+def scheming_isodatetime(field, schema):
+    def validator(key, data, errors, context):
+        value = data[key]
+        date_error = _('Date format incorrect')
+        time_error = _('Time format incorrect')
+        date = None
+
+        if isinstance(value, datetime.datetime):
+            return value
+        if not value is missing:
+            try:
+                 date = h.date_str_to_datetime(value)
+            except (TypeError, ValueError), e:
+                raise Invalid(date_error)
+        else:
+            extras = data.get(('__extras',))
+            if not extras or key[0] + '_date' not in extras:
+                if field.get('required'):
+                    not_empty(key, data, errors, context)
+            else:
+                for input_suffix in ['date', 'time']:
```

Comment:
if you've got a block of code you want to reuse you can create a nested function to call instead of a for loop with constant values. That will also save you the awkward `if input_suffix` checks below. Your nested functions have access to the locals in this function too.